Src/ILGPU/Runtime/ArrayViewExtensions.cs

```diff
@@ -0,0 +1,1072 @@
+        /// <remarks>This method is not supported on accelerators.</remarks>
+        [NotInsideKernel]
+        public static void CopyFromCPUAsync<T, TView>(
+            this TView target,
+            AcceleratorStream stream,
+            ref T cpuData,
+            long length)
+            where TView : IContiguousArrayView<T>
+            where T : unmanaged
+        {
+            if (length < 0)
+                throw new ArgumentOutOfRangeException(nameof(length));
+            if (length < 1)
+                return;
+
+            using var buffer = CPUMemoryBuffer.Create(ref cpuData, length);
+            target.Buffer.CopyFrom(
+                stream,
+                buffer.AsRawArrayView(),
+                target.GetIndexInBytes());
+        }
+
+        #endregion
+
+        #region Copy elements to/from CPU
+
+        /// <summary>
+        /// Copies from the source view into the given CPU target address while
+        /// synchronizing the current accelerator stream.
+        /// </summary>
+        /// <typeparam name="T">The element type.</typeparam>
+        /// <typeparam name="TView">The view type.</typeparam>
+        /// <param name="source">The source view instance.</param>
+        /// <param name="cpuData">The base address of the CPU buffer.</param>
+        /// <param name="length">The number of elements to copy.</param>
+        /// <remarks>This method is not supported on accelerators.</remarks>
+        [NotInsideKernel]
+        public static void CopyToCPUSync<T, TView>(
+            this TView source,
+            out T cpuData,
+            long length)
+            where TView : IContiguousArrayView<T>
+            where T : unmanaged =>
+            source.CopyToCPUSync(
+                source.GetDefaultStream(),
+                out cpuData,
+                length);
+
+        /// <summary>
+        /// Copies from the source view into the given CPU target address while
+        /// synchronizing the current accelerator stream.
+        /// </summary>
+        /// <typeparam name="T">The element type.</typeparam>
+        /// <typeparam name="TView">The view type.</typeparam>
+        /// <param name="source">The source view instance.</param>
+        /// <param name="stream">The used accelerator stream.</param>
+        /// <param name="cpuData">The base address of the CPU buffer.</param>
+        /// <param name="length">The number of elements to copy.</param>
+        /// <remarks>This method is not supported on accelerators.</remarks>
+        [NotInsideKernel]
+        public static void CopyToCPUSync<T, TView>(
+            this TView source,
+            AcceleratorStream stream,
+            out T cpuData,
+            long length)
+            where TView : IContiguousArrayView<T>
+            where T : unmanaged
+        {
+            source.CopyToCPUSync(stream, out cpuData, length);
+            stream?.Synchronize();
```

Comment:
QUESTION: Can `stream` ever be `null`? i.e. when would a caller pass `null`?

Reply:
The main "isuse" is that it is possible to create objects that are logically associated with a `CPUAccelerator` instance without explicitly instantiating a `CPUAccelerator` (e.g. an `UnmanagedMemoryBuffer`). Since there is no instance of a parent accelerator, there is no associated stream. What we could do is to use a static singleton instance of a `CPUStream` (since it doesn`t need to perform any special operations anyway) to avoid these null checks and to ensure that there are no edge cases where the `AcceleratorStream` is `null`.